zproject/backends.py

```diff
@@ -867,6 +888,62 @@ def sync_custom_profile_fields_from_ldap(
         except SyncUserException as e:
             raise ZulipLDAPException(str(e)) from e
 
+    def sync_groups_from_ldap(self, user_profile: UserProfile, ldap_user: _LDAPUser) -> None:
+        if not settings.LDAP_SYNC_GROUPS:
+            return
+
+        if user_profile.realm_id != settings.LDAP_GROUP_SYNC_REALM_ID:
+            return
+
+        try:
+            ldap_logger.debug("found groups for user: %s", str(user_profile))
+            ldap_group_name_set = set(ldap_user.group_names).intersection(set(settings.LDAP_GROUP_SYNC_LIST))
+
+            member_group_name_set = set(
+                [
+                    membership.user_group.name
+                    for membership in UserGroupMembership.objects.filter(
+                        user_group__realm=user_profile.realm,
+                        user_group__name__in=set(settings.LDAP_GROUP_SYNC_LIST),
+                        user_profile=user_profile,
+                    )
+                ]
+            )
+
+            ldap_logger.debug("ldap groups: %s", repr(ldap_group_name_set))
+            ldap_logger.debug("zulip groups: %s", repr(member_group_name_set))
+
+            add_group_list = []
+
+            for group_to_add in ldap_group_name_set.difference(member_group_name_set):
```

Comment:
```suggestion
            for group_to_add in UserGroup.objects.filter(name__in=ldap_group_name_set.difference(member_group_name_set), realm=user_profile.realm):
```
will fetch all the groups in bulk in one query, instead of the `group = UserGroup.objects.get(name=group_to_add, realm=user_profile.realm)` query in each iteration of the loop down below.

Reply:
see above, re-written

Reply:
@ClundXIII Hmm, you're still fetching in a loop with `.get()`s - my proposed change here would have fetched all the groups in a single query, which was the idea here. Same in the other similar loop below

Reply:
should be fixed now.